src/control/server/ctl_storage_rpc.go

```diff
@@ -242,68 +242,80 @@ func (c *ControlService) scmFormat(scmCfg storage.ScmConfig, reformat bool) (*ct
+		fmt.Sprintf(msgNvmeFormatSkip, srv.Index()))
 
 	c.log.Infof("formatting storage for %s instance %d (reformat: %t)",
-		DataPlaneName, i.Index(), reformat)
+		DataPlaneName, srv.Index(), reformat)
 
-	scmConfig := i.scmConfig()
+	scmConfig := srv.scmConfig()
 
-	// If not reformatting, check if SCM is already formatted.
-	if !reformat {
+	scmErrored := false
+	scmErr := func(err error) *ctlpb.ScmMountResult {
+		scmErrored = true
+		return newMntRet(c.log, "format", scmConfig.MountPoint,
+			ctlpb.ResponseStatus_CTL_ERR_SCM, err.Error(),
+			fault.ShowResolutionFor(err))
+	}
+
+	if srv.IsStarted() {
+		resp.Mrets = append(resp.Mrets, scmErr(errors.Errorf(
+			"instance %d: can't format storage of running instance", srvIdx)))
+	} else if !reformat {
+		// If not reformatting, check if SCM is already formatted.
 		var err error
-		needsScmFormat, err = i.NeedsScmFormat()
+		needsScmFormat, err = srv.NeedsScmFormat()
 		if err != nil {
-			return errors.Wrap(err, "unable to check storage formatting")
+			resp.Mrets = append(resp.Mrets, scmErr(err))
+		} else if !needsScmFormat {
+			resp.Mrets = append(resp.Mrets, scmErr(scm.FaultFormatNoReformat))
 		}
-		if !needsScmFormat {
-			err = scm.FaultFormatNoReformat
-			resp.Mrets = append(resp.Mrets,
-				newMntRet(c.log, "format", scmConfig.MountPoint,
-					ctlpb.ResponseStatus_CTL_ERR_SCM, err.Error(),
-					fault.ShowResolutionFor(err)))
-
-			if len(i.bdevConfig().DeviceList) > 0 {
-				resp.Crets = append(resp.Crets, skipNvmeResult)
-			}
+	}
 
-			return nil // don't continue if formatted and no reformat opt
+	if scmErrored {
+		if len(srv.bdevConfig().DeviceList) > 0 {
+			resp.Crets = append(resp.Crets, skipNvmeResult)
 		}
+		return // don't continue if we can't format SCM
 	}
 
 	// When SCM format is required, format and append to response results.
 	if needsScmFormat {
 		result, err := c.scmFormat(scmConfig, true)
 		if err != nil {
-			return errors.Wrap(err, "scm format") // return unexpected errors
-		}
-		resp.Mrets = append(resp.Mrets, result)
-
-		if result.State.Status != ctlpb.ResponseStatus_CTL_SUCCESS {
-			c.log.Error(msgFormatErr)
-			if len(i.bdevConfig().DeviceList) > 0 {
-				resp.Crets = append(resp.Crets, skipNvmeResult)
-			}
-
-			return nil // don't continue if we can't format SCM
+			resp.Mrets = append(resp.Mrets, scmErr(err))
+		} else {
+			resp.Mrets = append(resp.Mrets, result)
 		}
 	} else {
 		var err error
 		// If SCM was already formatted, verify if superblock exists.
-		needsSuperblock, err = i.NeedsSuperblock()
+		needsSuperblock, err = srv.NeedsSuperblock()
 		if err != nil {
-			return errors.Wrap(err, "unable to check instance superblock")
+			resp.Mrets = append(resp.Mrets, scmErr(err))
 		}
 	}
 
+	if scmErrored {
+		c.log.Errorf(msgFormatErr, srvIdx)
+		if len(srv.bdevConfig().DeviceList) > 0 {
+			resp.Crets = append(resp.Crets, skipNvmeResult)
+		}
+		return // don't continue if we can't format SCM
+	}
```

Comment:
Since we do something like this twice, would it be worth breaking out the SCM formatting into a helper function, and using defer to make sure this is done if we're exiting on error?

Reply:
done